moveit_ros/moveit_servo/include/moveit_servo/detail/input_check_valid.hpp

```diff
@@ -0,0 +1,78 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2021, PickNik Inc.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of PickNik Inc. nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/*      Title     : input_check_valid.hpp
+ *      Project   : moveit_servo
+ *      Created   : 11/06/2021
+ *      Author    : Tyler Weaver
+ */
+
+#pragma once
+
+#include <functional>
+#include <variant>
+#include <memory>
+#include <control_msgs/msg/joint_jog.hpp>
+#include <geometry_msgs/msg/twist_stamped.hpp>
+#include <moveit_servo/detail/input_command.hpp>
+#include <rclcpp/rclcpp.hpp>
+
+namespace moveit_servo
+{
+namespace detail
+{
+/**
+ * @brief      Visitor that checks if the input is valid, only calling next if that is true
+ */
+class InputCheckValid : public InputVisitor
+{
+  rclcpp::Node::SharedPtr node_ = nullptr;
+  std::string command_in_type_ = "unitless";
+  InputVisitor* next_ = nullptr;
```

Comment:
Why is this a raw pointer? Can it be a unique/shared pointer?

Reply:
They are raw pointers because all of these InputVisitor objects have their lifetime managed by ServoPipeline with a unique_ptr.  If I made them shared_ptr that would create a circular relationship:

If I made it a shared_ptr:
The lifetime of TimestampNow (next in the default case) would be tied to the lifetime of ServoPipeline 
The lifetime of TimestampNow would be tied to the lifetime of InputCheckValid
The lifetime of InputCheckValid would be tied to the lifetime of ServoPipeline

If I made it a unique_ptr:
The lifetime of TimestampNow would be tied to the lifetime of InputCheckValid
The lifetime of InputCheckValid would be tied to the lifetime of InputSubscriber
The lifetime of InputSubscriber would be tied to ServoPipeline (or whoever ended up controlling the InputSubscriber).

The second option would work, however, it makes testing more complex as I would need to move the unique_ptr<Visitor> into the constructor of each of these objects and retain a raw pointer to it.  Maybe that would be better?

In the current implementation, the lifetime of each of these is maintained by ServoPipeline (vector of unique_ptr<InputVisitor>) and raw pointers are handed to each one for the next one that it calls.

Do you think it would be better if I used unique_ptr instead of a non-owning raw pointer here?  Using the non-owning raw pointer does make the assumption that wherever the next object is in memory does not change and its lifetime is at least as long as the lifetime of the one we are constructing with the raw pointer.

Reply:
I think I might be wrong about how shared_ptr behaves, but I still don't think it is necessary or should be used here.

Reply:
I just tried turning these into unique_ptr and it breaks how I need to share access to ServoCalcs in servo.cpp.  I think these should stay raw pointers for now (potentially I could make them shared_ptr in the future) depending on how the rest of the ServoCalcs refactor goes.

Reply:
I've thought about this more and I changed the implementation to use no raw pointers.  It now uses shared_ptr which does work.  I'm going to delete my ramblings that actually don't make any sense.